src/parser.js

```diff
@@ -0,0 +1,104 @@
+'lang sweet.js';
+// @flow
+import { Either, Left } from './either';
+import { Maybe, Just, Nothing } from './maybe';
+
+type Mark = {};
+
+interface Context<T> {
+  next(): {
+    done: boolean;
+    value: T
+  };
+
+  mark(): Mark;
+
+  reset(Mark): void;
+}
+
+export class ImmutableContext<T> {
+  ctx: Context<T>;
+  value: Maybe<T>;
+
+  constructor(ctx: Context<T>) {
+    this.ctx = ctx;
+    let mark = ctx.mark();
+    let v = ctx.next();
+    ctx.reset(mark);
+    this.value = v.done ? new Nothing() : Maybe.of(v.value);
+  }
+
+  head() {
+    return this.value;
+  }
+
+  rest(): ImmutableContext<T> {
+    this.ctx.next();
+    return new ImmutableContext(this.ctx); 
+  }
+}
+
+export default class Parser<A, C> {
+  runner: ImmutableContext<C> => Either<string, [A, ImmutableContext<C>]>;
+
+  constructor(runner: ImmutableContext<C> => Either<string, [A, ImmutableContext<C>]>) {
+    this.runner = runner;
+  }
+  
+  run(ctx: ImmutableContext<*>): Either<string, [A, ImmutableContext<*>]> {
+    return this.runner(ctx);
+  }
+
+  static of(a: A): Parser<A, *> {
+    return new Parser(ctx => Either.of([a, ctx]));
+  }
+
+  static zero(msg: string = ''): Parser<*, *> {
```

Comment:
Not lawful as `zero` shouldn't take an argument. It might make sense to have `zero` call another method, `failure` (?) passing an empty string.